sqlmesh/core/macros.py

```diff
@@ -349,6 +354,65 @@ def engine_adapter(self) -> EngineAdapter:
             )
         return self.locals["engine_adapter"]
 
+    def _coerce(self, expr: exp.Expression, typ: t.Type) -> t.Any:
+        """Coerces the given expression to the specified type on a best-effort basis."""
+        try:
+            if typ is None:
+                return expr
+            base = t.get_origin(typ) or typ
+            if isinstance(expr, base):
+                return expr
+            if t.get_origin(typ) is t.Union:
```

Comment:
this is kind of concerning how complicated this is going to get with type coercision.

Reply:
Yeah, I think there is a line in the sand on what cases you want to handle. I think I drew the line in a pretty safe place. I wouldn't add more logic here personally. Its the users responsibility if they have some crazy marshaling to do. This handles all the basic cases I can think of with no further code which is extremely nice.

Reply:
I will add a dedicated test set for just this function so you can see it play out directly, instead of testing it indirectly via macros currently

Reply:
From there, we can assert that's all the cases we want to handle and anything else is user territory, which is already fine.